spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RetryGatewayFilterFactory.java

```diff
@@ -153,13 +161,56 @@ public void reset(ServerWebExchange exchange) {
 		exchange.getAttributes().remove(GATEWAY_ALREADY_ROUTED_ATTR);
 	}
 
+	public Mono<Void> preparedRequest(ServerWebExchange exchange,
+			GatewayFilterChain chain) {
+		return Mono.fromCallable(() -> {
+			Optional<DataBuffer> cachedRequestBuffer = exchange
+					.getAttributeOrDefault(RETRY_BODY_KEY, null);
+
+			Mono<Optional<DataBuffer>> preparedRequest;
+			// Place the original request buffer into the context as a single aggregated
+			// buffer
+			if (cachedRequestBuffer == null) {
+				preparedRequest = DataBufferUtils.join(exchange.getRequest().getBody())
+						.map(dataBuffer -> Optional.of(dataBuffer))
+						.defaultIfEmpty(Optional.empty())
+						.doOnNext(optionalDataBuffer -> exchange.getAttributes()
+								.put(RETRY_BODY_KEY, optionalDataBuffer));
+			}
+			else {
+				preparedRequest = Mono.just(cachedRequestBuffer);
+			}
+			return preparedRequest;
+		}).flatMap(
+				cachedRequestBuffer -> cachedRequestBuffer.flatMap(optionalDataBuffer -> {
+					ServerHttpRequestDecorator decorator = new ServerHttpRequestDecorator(
+							exchange.getRequest()) {
+						@Override
+						public Flux<DataBuffer> getBody() {
+							if (optionalDataBuffer.isPresent()) {
+								DataBuffer dataBuffer = optionalDataBuffer.get();
+								return Flux.defer(() -> {
+									// Retain the buffer, we will release down the chain
+									DataBufferUtils.retain(dataBuffer);
+									return Flux.just(dataBuffer.slice(0,
```

Comment:
Considered using wrap here on bytes, but this is possibly more performant.

Reply:
In https://github.com/spring-cloud/spring-cloud-gateway/commit/dc4181bbb0ad04383fb1722f589cfa0a199295e8#diff-0649883ef300daa40af407f414efa730R92 we moved to using `byte[]`.

I updated `getBody()` to the following and it seemed to work ok.
```
				public Flux<DataBuffer> getBody() {
					if (optionalDataBuffer.isPresent()) {
						DataBuffer dataBuffer = optionalDataBuffer.get();
						byte[] bytes = new byte[dataBuffer.readableByteCount()];
						dataBuffer.read(bytes);
						DataBufferUtils.release(dataBuffer);
						return Flux.defer(() -> {
							// Retain the buffer, we will release down the chain
							DataBufferUtils.retain(dataBuffer);
							DataBuffer buffer = exchange.getResponse().bufferFactory()
									.wrap(bytes);
							return Flux.just(buffer);
						});

					}
					return Flux.empty();
				}
```

Thoughts?

Reply:
Don’t retain after you release. Don’t defer on Unpooled buffer. 

The most recent Spring Framework commits appear to follow this convention.